src/perftest_resources.c

```diff
@@ -1517,6 +1531,44 @@ int create_reg_cqs(struct pingpong_context *ctx,
 		   struct perftest_parameters *user_param,
 		   int tx_buffer_depth, int need_recv_cq)
 {
+#ifdef HAVE_TD_API
+	struct ibv_cq_init_attr_ex send_cq_attr = {
+		.cqe = tx_buffer_depth * user_param->num_of_qps,
+		.cq_context = NULL,
+		.channel = ctx->send_channel,
+		.comp_vector = user_param->eq_num,
+	};
+
+	if (user_param->no_lock) {
+		send_cq_attr.parent_domain = ctx->pad;
+		send_cq_attr.comp_mask = IBV_CQ_INIT_ATTR_MASK_PD;
+	}
+
+	ctx->send_cq = ibv_cq_ex_to_cq(ibv_create_cq_ex(ctx->context, &send_cq_attr));
+	if (!ctx->send_cq) {
+		fprintf(stderr, "Couldn't create CQ\n");
+		return FAILURE;
+	}
+
+	if (need_recv_cq) {
+		struct ibv_cq_init_attr_ex recv_cq_attr = {
+			.cqe = user_param->rx_depth * user_param->num_of_qps,
+			.cq_context = NULL,
+			.channel = ctx->recv_channel,
+			.comp_vector = user_param->eq_num,
+		};
+		if (user_param->no_lock) {
+			recv_cq_attr.parent_domain = ctx->pad;
+			recv_cq_attr.comp_mask = IBV_CQ_INIT_ATTR_MASK_PD;
+		}
+		ctx->recv_cq = ibv_cq_ex_to_cq(ibv_create_cq_ex(ctx->context, &recv_cq_attr));
+		if (!ctx->recv_cq) {
+			fprintf(stderr, "Couldn't create a receiver CQ\n");
+			return FAILURE;
+		}
+	}
+	return SUCCESS;
+#else
```

Comment:
i think that you can make the creation of the cq's more generic by using the `ibv_cq_init_attr_ex` and `ibv_cq_ex_to_cq()` instead of 'ibv_create_cq()', and then minimize the filtering of the `HAVE_TD_API` to:
```
if (user_param->no_lock) {
		send_cq_attr.parent_domain = ctx->pad;
		send_cq_attr.comp_mask = IBV_CQ_INIT_ATTR_MASK_PD;
	}
```

Reply:
We try to ensure compatibility with older versions of rdma-core, where the cq_ex verbs may not be supported yet.
Since the td verbs were introduced later than the cq_ex verbs, we just use `HAVE_TD_API` here.

Reply:
I agree.
My suggestion would be to separate this into two `#ifdef `blocks: one for `HAVE_CQ_EX` and a nested one for `HAVE_TD_API`. This approach will provide greater flexibility for future additions to cq_ex features and make the code easier to review for developers who may not be familiar with TD.

Reply:
Sure. Thanks for the suggestion